### PR TITLE
Cache state to improve performance with URL-based storage.

### PR DESCRIPTION
**Pull request by divolgin**
What I Did
------------
Improved performance when using HTTP-based backend for state storage.

How I Did it
------------
State is now loaded and persisted once when using secret and URL-based backends.  File based storage is still updated continuously.  State is now served from an in-memory cache, and state manager is now implemented as a singleton.

How to verify it
------------
Run ship init and ship update.

Description for the Changelog
------------
Improved performance for HTTP-based state storage.

Picture of a Ship (not required but encouraged)
------------

### CODE REVIEW

File: pkg/ship/kustomize.go [hunk 1]
```diff
@@ -77,16 +81,16 @@ func (s *Ship) Init(ctx context.Context) error {
...
+	// if removeExistingState {
+	// 	if err := s.maybeWriteStateFromFile(); err != nil {
+	// 		return err
+	// 	}
+	// } else {
+	// 	debug.Log("event", "reset steps completed for existing state")
+	// 	if err := s.StateManager.ResetLifecycle(); err != nil {
+	// 		return errors.Wrap(err, "reset state.json completed lifecycle")
+	// 	}
+	// }
```
Comment (laverya) [inline pkg/ship/kustomize.go]: I'm not positive this can be removed
Does `ship update --headed` still work?

File: pkg/state/manager.go [hunk 2]
```diff
@@ -247,7 +308,11 @@ func (m *MManager) getStateSerializer() (stateSerializer, error) {
 
 	switch stateFrom {
 	case "file":
-		return newFileSerializer(m.FS, m.Logger), nil
+		// statePath := m.V.GetString("state-file")
+		// if statePath == "" {
+		// 	statePath = constants.StatePath
+		// }
```
Comment (laverya) [inline pkg/state/manager.go]: extra comments?

File: pkg/state/manager.go [hunk 3]
```diff
@@ -303,15 +380,17 @@ func (m *MManager) SaveKustomize(kustomize *Kustomize) error {
...
-	err := m.FS.RemoveAll(statePath)
-	if err != nil {
-		return errors.Wrap(err, "remove state file")
-	}
+	// err = s.Remove()
+	// if err != nil {
+	// 	return errors.Wrap(err, "remove state file")
+	// }
+
+	m.cachedState = &State{}
```
Comment (laverya) [inline pkg/state/manager.go]: Should this call the newly added `Remove` function too?

Also, commented code above

Comment (divolgin) [reply pkg/state/manager.go]: good catch. this was just for debugging.

**Pull request merged**